src/envs/pddl_env.py

```diff
@@ -0,0 +1,442 @@
+"""Environments that are derived from PDDL.
+
+There are no continuous aspects of the state or action space. These
+environments are similar to PDDLGym.
+"""
+
+import abc
+from typing import Callable, Collection, Dict, List, Optional, Sequence, Set, \
+    Tuple, cast
+
+import numpy as np
+from gym.spaces import Box
+from pyperplan.pddl.parser import TraversePDDLDomain, TraversePDDLProblem, \
+    parse_domain_def, parse_lisp_iterator, parse_problem_def
+
+from predicators.src import utils
+from predicators.src.envs import BaseEnv
+from predicators.src.envs.pddl_procedural_generation import \
+    create_blocks_pddl_generator
+from predicators.src.settings import CFG
+from predicators.src.structs import Action, Array, GroundAtom, Image, \
+    LiftedAtom, Object, ParameterizedOption, PDDLEnvSimulatorState, \
+    PDDLProblemGenerator, Predicate, State, STRIPSOperator, Task, Type, \
+    Variable, _GroundSTRIPSOperator
+
+###############################################################################
+#                                    ABCs                                     #
+###############################################################################
+
+
+class _PDDLEnv(BaseEnv):
+    """An environment induced by PDDL.
+
+    The state space is mostly unused. The continuous vectors are dummies. What
+    is actually used is state.simulator_state, which holds the current ground
+    atoms (see PDDLEnvSimulatorState). Note that we need to use this pattern,
+    as opposed to just maintaining the ground atoms internally in the env,
+    because the predicate classifiers need access to the ground atoms.
+
+    The action space is hacked to conform to our convention that actions
+    are fixed dimensional vectors. Users of this class should not need
+    to worry about the action space because it would never make sense to
+    learn anything using this action space. The dimensionality is 1 +
+    max operator arity. The first dimension encodes the operator. The
+    next dimensions encode the object used to ground the operator. The
```

Comment:
but doesn't this require knowing the number of objects? (or at least an upper bound on it)

Reply:
it only requires knowing the max operator arity

Reply:
oh yeah that's cuz your action space has infs on the upper bound